[DOC]Restructuring (#3890)
pachyderm/pachyderm: doc/concepts/pipeline-concepts/spout.md

# Spout

> A spout is a type of pipeline that processes

Reviewer: ```suggestion
A spout is a type of pipeline that ingests
```

> streaming data that is continuously generated from

Reviewer: ```suggestion
streaming data.
```

> multiple sources. Generally, you use spouts for

Reviewer: ```suggestion
Generally, you use spouts for
```

situations when the interval between new data generation

> is large, but the latency requirement to start the

Reviewer: ```suggestion
is large or sporadic, but the latency requirement to start the
```

processing is short. Therefore, a regular pipeline

> with periodic processing might not be an optimal solution.

Reviewer: ```suggestion
with a cron input that polls for new data might not be an optimal solution.
```

Examples of streaming data include a message queue,
a database transactions log, event notifications,
and others. In spouts, your code runs continuously and writes the

> results to the pipeline's output location `pfs/out`.

Reviewer: ```suggestion
results to the pipeline's output location, `pfs/out`.
```

Every time you create a complete `.tar` archive,
Pachyderm creates a new commit and triggers the pipeline to
process it.

One main difference from regular pipelines is that
you cannot specify input in the spout pipeline.

Another important aspect is that in spouts, `pfs/out` is
a *named pipeline*, or *First in, First Out* (FIFO), and is not
a directory like in standard pipelines. Unlike

> the traditional and familiar to most Linux users

Author: Rephrased.

pipe, a *named pipe* enables two system processes to access
the pipe simultaneously and gives one of the processes read-only and the other
process write-only access. Therefore, the two processes can simultaneously
read and write to the same pipe.

To create a spout pipeline, you need the following items:

> * A source of streaming data

Author: This list consists of phrases, not complete sentences with a subject and predicate. In general, phrases do not require periods.

* Docker container with your spout code that reads from the data source
* A spout pipeline specification file that uses the container

> A spout code performs the following actions:

Reviewer: ```suggestion
Your spout code performs the following actions:
```

> 1. Connects to the specified streaming data source.

Reviewer 2: All 1. again?

Author: yup =)

1. Reads the data from the streaming data source.
1. Packages the data into a `tar` stream.
1. Writes the `tar` stream into the `pfs/out` pipe.

> 

Reviewer 3: I might put it this way
1. Connects to the specified streaming data source.
1. Opens /pfs/out as a named pipe
1. Reads the data from the streaming data source.
1. Packages the data into a `tar` stream.
1. Writes the `tar` stream into the `pfs/out` pipe, being careful to retry in case of transient errors produced by closing a previous write to the pipe
1. Closes the `tar` stream and connection to `/pfs/out`, which produces the commit.

> 1. Writes the `tar` stream into the `pfs/out` pipeline.

Reviewer: ```suggestion
1. Writes the `tar` stream into the `pfs/out` pipe.
```

A minimum spout specification must include the following
parameters:

| Parameter   | Description |
| ----------- | ----------- |
| `name`      | The name of your data pipeline and the output repository. You can set an <br> arbitrary name that is meaningful to the code you want to run. |
| `transform` | Specifies the code that you want to run against your data, such as a Python <br> or Go script. Also, specifies a Docker image that you want to use to run that script. |

> | `overwrite` | Specifies whether to overwrite the existing content of the file <br> from previous commits or previous calls to the `put file` command <br> within this commit. The default value is `false`. |

Reviewer: is overwrite required? If not, then it shouldn't be in the "minimum spout specification". We still should include it in this doc though. Or just take out the word minimum :)

Author: It is an optional parameter. I added `(Optional)`. I want to keep it in the table though.
